update llama4 chat template and pythonic parser

upfixer opened the pull request:
update llama4 chat template and pythonic parser to remove <|python_start|> and <|python_end|>

<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

- llama4 chat template updated in https://www.llama.com/docs/model-cards-and-prompt-formats/llama4/ and vllm https://github.com/vllm-project/vllm/pull/17917 

- llama4 model [issue](https://github.com/vllm-project/vllm/issues/16214), it outputs unexpected <|python_start|> and <|python_end|>




## Modifications

- updated llama4 chat template 
- updated pythonic parser to remove <|python_start|> and <|python_end|>

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

upfixer committed:
update llama4 chat template and pythonic parser to remove <|python_start> and <|python_end|>

upfixer committed:
merge main

CatherineSue commented on python/sglang/srt/function_call/pythonic_detector.py:
```diff
@@ -32,13 +32,24 @@ def __init__(self):
+
     def has_tool_call(self, text: str) -> bool:
-        return bool(self.tool_call_regex.search(text.strip()))
+        return bool(self.tool_call_regex.match(self._text_strip(text.strip())))
```
We should not use match here. This will make it failed to check if there is a mixed text with the function call.

`match` vs `search`
`match()`: Only matches if the regex pattern matches from the very beginning of the string
`search()`: Finds the pattern anywhere in the string

CatherineSue commented on python/sglang/srt/function_call/pythonic_detector.py:
```diff
@@ -125,6 +136,9 @@ def parse_streaming_increment(
         then parses and emits any detected calls.
         """
+        # remove unexpected <|python_start|> and <|python_end|> for llama4
+        new_text = self._text_strip(new_text)
```
Is there a pattern for when <|python_start|> or <|python_end|> will happen? Will it always be a single token?

`new_text` will always be one token. It will fail to match if the new_text is like `<|pyhton_`, then the next one is `start|>`

zhyncs committed:
Merge branch 'main' into sglang/llama4-tool-call-update

CatherineSue commented:
> ```
> ============================================================
> Testing NON-STREAMING: Tool choice AUTO
> tool_choice: auto
> ============================================================
> ✅ Request successful
> Content: To answer the question about the weather in the top 5 most populated cities in the US, I first need to identify what those cities are. Then, I can use the `get_weather` tool to find the current weather for each city.
> 
> The top 5 most populated cities in the US are:
> 1. New York City, NY
> 2. Los Angeles, CA
> 3. Chicago, IL
> 4. Houston, TX
> 5. Phoenix, AZ
> 
> Next, I will use the `get_weather` tool for each of these cities.
> 
> [get_weather(city="New York City"), get_weather(city="Los Angeles"), get_weather(city="Chicago"), get_weather(city="Houston"), get_weather(city="Phoenix")]
> Tool calls: None
> ℹ️  INFO: Auto mode - tool calls: False
> ```


This is related to the comment I left above regarding `match` vs `search`. Please address them and retest this one. Thanks!

CatherineSue committed:
Resolve comments

- Updated `has_tool_call` method to use `search` instead of `match` for regex.
- Introduced `_strip_and_split_buffer` method to manage special tokens and partial tokens.
- Enhanced `parse_streaming_increment` to correctly handle and retain special tokens around tool calls.
- Improved unit tests for parsing with special tokens to ensure accurate detection and parameter extraction.

CatherineSue commented:
Updated result:
```
============================================================
Testing NON-STREAMING: Tool choice AUTO
tool_choice: auto
============================================================
✅ Request successful
Content: To determine the weather in the top 5 most populated cities in the US, we first need to identify those cities and then retrieve their current weather information.

The top 5 most populated cities in the US are typically New York City, Los Angeles, Chicago, Houston, and Phoenix.

1. **Identify the top 5 cities**: 
   - New York City
   - Los Angeles
   - Chicago
   - Houston
   - Phoenix

2. **Get the weather for each city**:
   - We need to use the `get_weather` function for each of these cities.

Let's start by getting the weather for the first city, New York City.


Tool calls: [ChatCompletionMessageToolCall(id='call_rX_4PJH6RJWfOMmvnw6wDw', function=Function(arguments='{"decision": "TOOL", "content": "Get the weather for New York City using get_weather function"}', name='make_next_step_decision'), type='function', index=None), ChatCompletionMessageToolCall(id='call_xeiv4bY-TfCh5mbZqH2-UA', function=Function(arguments='{"city": "New York City"}', name='get_weather'), type='function', index=None)]
ℹ️  INFO: Auto mode - tool calls: True
```

Tool calls successfully detected

zhyncs committed:
Merge branch 'main' into sglang/llama4-tool-call-update
